_deploy/r/gnoswap/gnft/gnft_test.gno

```diff
@@ -0,0 +1,450 @@
+			tokenIdToBurn: 99,
+			shouldPanic:   true,
+			panicMsg:      errInvalidTokenId,
+		},
+		{
+			name:          "burn token id(2)",
+			callerRealm:   std.NewCodeRealm(consts.POSITION_PATH),
+			tokenIdToBurn: 2,
+			shouldPanic:   false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			std.TestSetRealm(tt.callerRealm)
+
+			if tt.shouldPanic {
+				uassert.PanicsWithMessage(t, tt.panicMsg, func() {
+					Burn(tid(tt.tokenIdToBurn))
+				})
+			} else {
+				uassert.NotPanics(t, func() {
+					Burn(tid(tt.tokenIdToBurn))
+				})
+			}
+		})
+	}
+}
+
+func TestSetTokenURI(t *testing.T) {
+	tests := []struct {
+		name        string
+		callerRealm std.Realm
+		tokenId     uint64
+		shouldErr   bool
+		errMsg      string
+	}{
+		{
+			name:      "set token uri without permission",
+			tokenId:   1,
+			shouldErr: true,
+			errMsg:    "caller is not token owner",
+		},
+		{
+			name:      "set token uri of non-minted token id",
+			tokenId:   99,
+			shouldErr: true,
+			errMsg:    errInvalidTokenId,
+		},
+		{
+			name:        "set token uri of token id(1)",
+			callerRealm: addr01Realm,
+			tokenId:     1,
+			shouldErr:   false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			std.TestSetRealm(tt.callerRealm)
+			_, err := SetTokenURI(tid(tt.tokenId))
+
+			if tt.shouldErr {
+				uassert.ErrorContains(t, err, tt.errMsg)
+			} else {
+				uassert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestTokenURI(t *testing.T) {
+	resetObject(t)
+
+	tests := []struct {
+		name        string
+		setup       func()
+		tokenId     uint64
+		shouldPanic bool
+		panicMsg    string
+	}{
+		{
+			name:        "get token uri of non-minted token id",
+			tokenId:     99,
+			shouldPanic: true,
+			panicMsg:    errInvalidTokenId,
+		},
+		{
+			name: "get token uri of minted token but not set token uri",
+			setup: func() {
+				std.TestSetRealm(positionRealm)
+				Mint(a2u(addr01), tid(1))
+			},
+			tokenId:     1,
+			shouldPanic: true,
+			panicMsg:    errInvalidTokenId,
+		},
+		{
+			name: "get token uri of minted token after setting token uri",
+			setup: func() {
+				std.TestSetRealm(addr01Realm)
+				SetTokenURI(tid(1))
+			},
+			tokenId: 1,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup()
+			}
+
+			if tt.shouldPanic {
+				uassert.PanicsWithMessage(t, tt.panicMsg, func() {
+					TokenURI(tid(tt.tokenId))
+				})
+			} else {
+				uassert.NotEmpty(t, TokenURI(tid(tt.tokenId)))
+			}
+		})
+	}
+}
+
+func TestSetTokenURILast(t *testing.T) {
+	resetObject(t)
+	std.TestSetRealm(positionRealm)
+	Mint(a2u(addr01), tid(1))
+	Mint(a2u(addr01), tid(2)) // last minted
+
+	t.Run("set token uri last", func(t *testing.T) {
+		std.TestSetRealm(addr01Realm)
+		SetTokenURILast()
+	})
+
+	t.Run("token uri(2)", func(t *testing.T) {
+		uassert.NotEmpty(t, TokenURI(tid(2)))
+	})
+}
+
```

Comment:
We need to make sure that the generateRandInstance function is randomized.

Reply:
In particular, you should consider whether the time values will not be the same for transaction processing in the same block.

Reply:
https://github.com/gnoswap-labs/gnoswap/pull/443#discussion_r1893377900

you already fixed it.